x-pack/plugin/search-business-rules/src/main/java/org/elasticsearch/xpack/searchbusinessrules/PinnedQueryBuilder.java

```diff
@@ -0,0 +1,218 @@
+ */
+
+package org.elasticsearch.xpack.searchbusinessrules;
+
+import org.apache.lucene.search.BooleanClause;
+import org.apache.lucene.search.BooleanQuery;
+import org.apache.lucene.search.BoostQuery;
+import org.apache.lucene.search.CappedScoreQuery;
+import org.apache.lucene.search.ConstantScoreQuery;
+import org.apache.lucene.search.DisjunctionMaxQuery;
+import org.apache.lucene.search.MatchNoDocsQuery;
+import org.apache.lucene.search.Query;
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.ParsingException;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.lucene.search.Queries;
+import org.elasticsearch.common.xcontent.ObjectParser;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.index.mapper.IdFieldMapper;
+import org.elasticsearch.index.mapper.MappedFieldType;
+import org.elasticsearch.index.query.AbstractQueryBuilder;
+import org.elasticsearch.index.query.MatchNoneQueryBuilder;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.QueryRewriteContext;
+import org.elasticsearch.index.query.QueryShardContext;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+
+import static org.elasticsearch.common.xcontent.ObjectParser.fromList;
+
+/**
+ * A query that will promote selected documents (identified by ID) above matches produced by an "organic" query. In practice, some upstream
+ * system will identify the promotions associated with a user's query string and use this object to ensure these are "pinned" to the top of
+ * the other search results.
+ */
+public class PinnedQueryBuilder extends AbstractQueryBuilder<PinnedQueryBuilder> {
+    public static final String NAME = "pinned";
+
+    private static final ParseField IDS_FIELD = new ParseField("ids");
+    public static final ParseField ORGANIC_QUERY_FIELD = new ParseField("organic");
+
+    private final Set<String> ids = new LinkedHashSet<>();
+    private QueryBuilder organicQuery;
+
+    // Organic queries will have their scores capped to this number range,
+    private static final float MAX_ORGANIC_SCORE = Float.intBitsToFloat((0xfe << 23)) - 1;
+
+    /**
+     * Creates a new PinnedQueryBuilder
+     */
+    public PinnedQueryBuilder() {
+        // nothing to do
+    }
+
+    /**
+     * Read from a stream.
+     */
+    public PinnedQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        Collections.addAll(ids, in.readStringArray());
+        organicQuery = in.readNamedWriteable(QueryBuilder.class);
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeStringArray(ids.toArray(new String[ids.size()]));
```

Comment:
you can do `out.writeCollection(ids, out::writeString)` here and `in.readCollection(in::readString, sz -> new LinkedHashSet<>())` in the constructor that takes a string to avoid creating a temporary array.

Reply:
There's a `writeStringCollection()` for the out but the read doesn't compile (it expects the `readString` method to take a StreamInput arg)

Reply:
oh sorry, it should have been `in.readCollection(StreamInput::readString, sz -> new LinkedHashSet<>())`

Reply:
Thanks, that readCollection method is currently private - ok to change visibility?

Reply:
Fine with me, but maybe we don't even need it if we move to a List (see my other comment about equals).